File: javascript/data-structures/p233_Stack.js
/*
    CLRS Section 10.1, p. 233
*/

class Stack {
    constructor(S = [], top = -1) {
        this.S = S;
        this.top = top;
    }

    stackEmpty = () => this.top === -1;

    push = (x) => {
        this.top++;
        this.S[this.top] = x;
    }
    
    pop = () => {
        if (this.stackEmpty()) {
            throw new Error('underflow');
        } else {
            this.top--;
            return this.S.splice(this.top + 1, 1, undefined)[0];
        }
    }
}

module.exports = {
    Stack
}

File: python/data_structures/p233_stack.py
'''
    CLRS Section 10.1, p. 233
'''

class Stack:
    def __init__(self, S=None, top=-1):
        if S is None:
            self.S = []
        else:
            self.S = S
        self.top = top
    
    def stack_empty(self):
        return self.top == -1
    
    def push(self, x):
        self.top += 1
        self.S[self.top] = x
    
    def pop(self):
        if self.stack_empty():
            raise Exception('underflow')
        else:
            self.top -= 1
            self.S.append(None)
            return self.S.pop(self.top + 1)
